src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -229,152 +299,653 @@ ValidatorList::CacheValidatorFile(
+        smallMsg->set_manifest(largeMsg.manifest());
+
+        for (std::size_t i = begin; i < end; ++i)
+        {
+            *smallMsg->add_blobs() = largeMsg.blobs(i);
+        }
+        auto message = std::make_shared<Message>(
+            *smallMsg, protocol::mtVALIDATORLISTCOLLECTION);
+
+        if (message->getBuffer(compression::Compressed::Off).size() > maxSize)
+        {
+            // free up the message space
+            message.reset();
+            smallMsg.reset();
+            return splitMessage(messages, largeMsg, maxSize, begin, end);
+        }
+        else
+        {
+            messages.emplace_back(
+                message, sha512Half(*smallMsg), smallMsg->blobs_size());
+            return messages.back().numVLs;
+        }
+    }
+    return 0;
+}
+
+// Build a v1 protocol message using only the current VL
+std::size_t
+buildValidatorListMessage(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    std::uint32_t rawVersion,
+    std::string const& rawManifest,
+    ValidatorBlobInfo const& currentBlob,
+    std::size_t maxSize)
+{
+    assert(messages.empty());
+    protocol::TMValidatorList msg;
+    auto const manifest =
+        currentBlob.manifest ? *currentBlob.manifest : rawManifest;
+    auto const version = 1;
+    msg.set_manifest(manifest);
+    msg.set_blob(currentBlob.blob);
+    msg.set_signature(currentBlob.signature);
+    // Override the version
+    msg.set_version(version);
+
+    auto message = std::make_shared<Message>(msg, protocol::mtVALIDATORLIST);
+    if (message->getBuffer(compression::Compressed::Off).size() <= maxSize)
+    {
+        messages.emplace_back(message, sha512Half(msg), 1);
+        return 1;
+    }
+    return 0;
+}
+
+// Build a v2 protocol message using all the VLs with sequence larger than the
+// peer's
+std::size_t
+buildValidatorListMessage(
+    std::vector<ValidatorList::MessageWithHash>& messages,
+    std::uint64_t peerSequence,
+    std::uint32_t rawVersion,
+    std::string const& rawManifest,
+    std::map<std::size_t, ValidatorBlobInfo> const& blobInfos,
+    std::size_t maxSize)
+{
+    assert(messages.empty());
+    protocol::TMValidatorListCollection msg;
+    auto const version = rawVersion < 2 ? 2 : rawVersion;
+    msg.set_version(version);
+    msg.set_manifest(rawManifest);
+
+    for (auto const& [sequence, blobInfo] : blobInfos)
+    {
+        if (sequence <= peerSequence)
+            continue;
+        protocol::ValidatorBlobInfo& blob = *msg.add_blobs();
+        blob.set_blob(blobInfo.blob);
+        blob.set_signature(blobInfo.signature);
+        if (blobInfo.manifest)
+            blob.set_manifest(*blobInfo.manifest);
+    }
+    assert(msg.blobs_size() > 0);
+    auto message =
+        std::make_shared<Message>(msg, protocol::mtVALIDATORLISTCOLLECTION);
+    if (message->getBuffer(compression::Compressed::Off).size() > maxSize)
+    {
+        // throw away the Message and split into smaller messages
```

Comment:
Seems like you can avoid creating `message` in this case by comparing the sum of `msg.ByteSize()` and `compression::headerBytes` with `maxSize` and creating `message` only if it will be sent as a single message.

Reply:
In order to conveniently get the message size you can add a static function to the `Message` class that encapsulates [this logic](https://github.com/ximinez/rippled/blob/6ebf80b80734867cb7781d021838b88f28988f17/src/ripple/overlay/impl/Message.cpp#L35).

Reply:
Great tip! I didn't dig enough into it to figure out how it was computed. I assumed it was way more complicated than that.